### DISCUSSION TITLE
Enable cross-controller communication with events?

### DISCUSSION THREAD

**Discussion by jlw** [josefarias/hotwire_combobox#71] [Ideas]
I am building an app where I have a "required-submit" controller that A) highlights required fields only when they are empty and B) disables the submit button until all required fields have a value (it's not doing any other validation at the moment).

In order to make this work with this gem's elements, the only workable solution I've come across so far is to create my own event listener for each hidden input in my controller.

Since stimulus has it's own model for [cross-controller coordination](https://stimulus.hotwired.dev/reference/controllers#cross-controller-coordination-with-events), I'm wondering if there's a _clean_ way to add a helper argument to tell the combobox controller our desired recipient controller + action for routing a new event dispatch.

Is this an edge case requirement? I don't feel confident enough that this is a common case to start working on an implementation just yet, but am willing to change my mind based on feedback here.

**Comment by josefarias** [comment]
Thanks @jlw. Yeah, I think we can emit custom events when the combobox locks in a selection and when it becomes valid/invalid. It will then be trivial to hook it into other controllers once https://github.com/josefarias/hotwire_combobox/discussions/36 is implemented. Something like:

```erb
<%= combobox_tag :foo, @options do |combobox| %>
  <% combobox.hidden_input_attrs data: { action: "hw-combobox:invalid->form#doSomething" } %>
<% end %>
```

Don't love the tight coupling to `hidden_input` as some users won't know that that means.

And I'm unsure how the selection event should behave on freetext inputs.

But both of those should be addresable.

Related https://github.com/josefarias/hotwire_combobox/issues/24

**Reply by josefarias** [reply]
Actually, I forgot any passed data attrs are merged into the fieldset data attrs. This is different from most other unrecognized attrs, which are merged into the input's attrs.

So using the custom events could be as simple as:

`<%= combobox_tag :foo, @options, data: { action: "hw-combobox:invalid->form#doSomething" } %>`

Provided our custom events bubble up.

**Comment by josefarias** [comment]
https://github.com/josefarias/hotwire_combobox/pull/75

**Comment by josefarias** [comment]
Merged!
